File: 2019/5.hs
import qualified Debug.Trace as DT

csplit str = aux str "" [] where
  aux (',':rest) word words = aux rest "" $ words ++ [word]
  aux (ch:rest) word words = aux rest (word ++ [ch]) words
  aux [] word words = words ++ [word]

input :: IO [Int]
input = fmap (\s -> map read (csplit s)) (readFile "5.input")

tape_set :: [Int] -> Int -> Int -> [Int]
tape_set tape dest v = let (l,r) = splitAt dest tape in l ++ [v] ++ (tail r)

tape_read tape pos mode = let read = tape !! pos
                              in case mode of
                                   0 -> tape !! read
                                   1 -> read

tape_arithmetic op tape op_pos modes = tape_set tape dest val
  where a = tape_read tape (op_pos + 1) (modes !! 0)
        b = tape_read tape (op_pos + 2) (modes !! 1)
        val = op a b
        dest = tape !! (op_pos + 3)

tape_add = tape_arithmetic (+)
tape_mul = tape_arithmetic (*)
tape_less_than = tape_arithmetic (\a b -> if a < b then 1 else 0)
tape_equals = tape_arithmetic (\a b -> if a == b then 1 else 0)

-- tape_write :: [Int] -> Int -> Int -> [Int]
tape_write tape op_pos val = tape_set tape dest val
  where dest = tape !! (op_pos + 1)

jump_if pred tape pos modes = let val = tape_read tape (pos + 1) (modes !! 0)
                                  jump = pred val
                              in case jump of
                                True -> Just (tape_read tape (pos + 2) (modes !! 1))
                                False -> Just (pos + 3)

jump_ifnz = jump_if (\n -> n /= 0)
jump_ifz  = jump_if (\n -> n == 0)

decode code = let op = mod code 100
                  c = mod (quot code 100) 10
                  b = mod (quot code 1000) 10
              in (op, [c, b])

step tape pos input output = let (op, modes) = decode (tape !! pos)
  in case op of
       99 -> (tape, input, output, Nothing) -- halt
       1 -> (tape_add tape pos modes, input, output, Just (pos + 4))
       2 -> (tape_mul tape pos modes, input, output, Just (pos + 4))
       3 -> (tape_write tape pos (head input), tail input, output, Just (pos + 2))
       4 -> (tape, input, output ++ [tape_read tape (pos+1) (head modes)], Just (pos + 2))
       5 -> (tape, input, output, jump_ifnz tape pos modes)
       6 -> (tape, input, output, jump_ifz tape pos modes)
       7 -> (tape_less_than tape pos modes, input, output, Just (pos + 4))
       8 -> (tape_equals tape pos modes, input, output, Just (pos + 4))

run tape pos input output = let (state, inp, out, next) = step tape pos input output in
                              case next of
                                Just op -> run state op inp out
                                Nothing -> (state, out)

part1 = do
  s <- input
  let z = run s 0 [1] []
  return z

part2 = do
  s <- input
  let z = run s 0 [5] []
  return z

File: 2019/5.py
from dataclasses import dataclass


@dataclass
class intcode:
    tape: list
    pos: int
    input: list
    output: list


def decode_op(instruction):
    op = instruction % 100
    instruction //= 100
    modes = [0] * 3
    argument_index = 0
    while instruction > 0:
        mode = instruction % 10
        modes[argument_index] = mode
        instruction //= 10
        argument_index += 1
    return op, modes


def tape_read(tape, source, mode):
    source = tape[source]
    if mode == 0:
        source = tape[source]
    return source


def op_arithmetic(op):
    def interpret(state, modes):
        A = tape_read(state.tape, state.pos + 1, modes[0])
        B = tape_read(state.tape, state.pos + 2, modes[1])
        dest = state.tape[state.pos + 3]
        val = op(A, B)
        state.tape[dest] = val
        state.pos += 4
        return state

    return interpret


def op_read(state, modes):
    value = state.input[0]
    state.input.pop(0)
    dest = state.tape[state.pos + 1]
    state.tape[dest] = value
    state.pos += 2
    return state


def op_write(state, modes):
    value = tape_read(state.tape, state.pos + 1, modes[0])
    state.output.append(value)
    state.pos += 2
    return state


def op_jnz(state, modes):
    pred = tape_read(state.tape, state.pos + 1, modes[0])
    if pred != 0:
        dest = tape_read(state.tape, state.pos + 2, modes[1])
        state.pos = dest
    else:
        state.pos += 3
    return state


def op_jz(state, modes):
    pred = tape_read(state.tape, state.pos + 1, modes[0])
    if pred == 0:
        dest = tape_read(state.tape, state.pos + 2, modes[1])
        state.pos = dest
    else:
        state.pos += 3
    return state


def op_halt(state, modes):
    state.pos = None
    return state


instructions = {
    1: op_arithmetic(lambda a, b: a + b),
    2: op_arithmetic(lambda a, b: a * b),
    3: op_read,
    4: op_write,
    5: op_jnz,
    6: op_jz,
    7: op_arithmetic(lambda a, b: 1 if a < b else 0),
    8: op_arithmetic(lambda a, b: 1 if a == b else 0),
    99: op_halt,
}


def run(tape, input):
    state = intcode(tape.copy(), 0, input, [])

    while state.pos is not None:
        op, modes = decode_op(state.tape[state.pos])
        state = instructions[op](state, modes)
    return state


t1 = [3, 3, 1105, -1, 9, 1101, 0, 0, 12, 4, 12, 99, 1]
t2 = [3, 9, 8, 9, 10, 9, 4, 9, 99, -1, 8]
t3 = [3, 3, 1108, -1, 8, 3, 4, 3, 99]
t4 = [3, 3, 1107, -1, 8, 3, 4, 3, 99]
t5 = [3, 12, 6, 12, 15, 1, 13, 14, 13, 4, 13, 99, -1, 0, 1, 9]
t6 = [3, 3, 1105, -1, 9, 1101, 0, 0, 12, 4, 12, 99, 1]
t7 = [
    3, 21, 1008, 21, 8, 20, 1005, 20, 22, 107, 8, 21, 20, 1006, 20, 31, 1106, 0,
    36, 98, 0, 0, 1002, 21, 125, 20, 4, 20, 1105, 1, 46, 104, 999, 1105, 1, 46,
    1101, 1000, 1, 20, 4, 20, 1105, 1, 46, 98, 99
]

tape = [int(n) for n in open("5.input").read().split(",")]


def part1(tape):
    tape = tape.copy()
    return run(tape, [1]).output


def part2(tape):
    tape = tape.copy()
    return run(tape, [5]).output
